### DISCUSSION TITLE
`vim.ui_attach`: how to display prompt text in fast event?

### DISCUSSION THREAD

**Discussion by przepompownia** [neovim/neovim#31522] [Q&A] [answered]
```lua
local api = vim.api
local ns = api.nvim_create_namespace('messageRedirection')

local function log(msg)
  if 0 == #vim.trim(msg) then
    return
  end
  vim.schedule(function ()
    api.nvim_buf_set_lines(1, -1, -1, true, vim.split(msg, '\n'))
    vim.cmd.normal 'G'
  end)
end

local function redraw()
  vim.schedule(function ()
    vim.cmd.redraw()
  end)
end

api.nvim_create_autocmd('CmdlineEnter', {
  callback = function ()
    vim.ui_detach(ns)
    redraw()
  end
})

local function attach()
  vim.ui_attach(ns, {ext_messages = true, ext_cmdline = false}, function (event, kind, content, _replace_last)
    if event == 'msg_show' then
      log(vim.inspect(content))
    end
    redraw()
  end)
end

api.nvim_create_autocmd({'CmdlineLeave'}, {callback = attach})

api.nvim_create_autocmd({'UIEnter'}, {
  callback = function ()
    attach()
    vim.fn.inputlist({'Select: ', 1, 2})
  end
})
```
Let's try the above example. How to display the prompt from `inputlist` on time (i.e. without `vim.schedule`)? The second question is how to avoid being stuck on a blank screen while waiting for user input. 

I haven't bisected Noice yet to check how it behaves in such cases.

**Comment by luukvbaal** [comment, marked as answer]
The idea is you schedule messages and put them in a queue. Then if a prompt happens and the queue is non-empty, you emit the queued messages as well. Using this in my (local) version of the default ext UI:
```lua
vim.ui_attach(ext.ns, { ext_cmdline = true, ext_messages = true }, function(event, ...)
  if vim.in_fast_event() then
    if event:match('msg_') then
      ext.msg.queue[#ext.msg.queue + 1] = { event, ... }
    end
    scheduled_ui_callback(event, ...)
  else
    ui_callback(event, ...)
  end
end)
```

> The second question is how to avoid being stuck on a blank screen while waiting for user input.

If you mean how do we display the user input; that's not possible currently. The input is emitted as `msg_show`, I think ideally it would be emitted as `cmdline_show`, or otherwise a non-fast `msg_show` kind (https://github.com/folke/noice.nvim/issues/1002#issuecomment-2495419639).

**Reply by przepompownia** [reply]
I suspect that consuming that queue along with the nearest message incoming from non-fast event is too late (at least in the same way as before). 

I see that "fast" `list_cmd` message appears first and it could be consumed from the queue along with `number_prompt` message that comes later already in non-fast context. 

```lua
{
  c = { { 0, "Select: \n1\n2\n", 0 } },
  e = "msg_show",
  fast = true,
  k = "list_cmd",
  r = "none",
  time = 271068501.3056
}
{
  c = { { 0, "Type number and <Enter> or click with the mouse (q or empty cancels): ", 0 } },
  e = "msg_show",
  fast = false,
  k = "number_prompt",
  r = "none",
  time = 271068501.32236
}
```
Can be `ext_cmdline` at all useful in this case? 

Modified `init.lua` with still useless queue and still with unused `ext_cmdline`:
```lua
local api = vim.api
local ns = api.nvim_create_namespace('messageRedirection')

local function log(msg)
  if 0 == #vim.trim(msg) then
    return
  end
  vim.schedule(function ()
    api.nvim_buf_set_lines(1, -1, -1, true, vim.split(msg, '\n'))
    vim.cmd.normal 'G'
  end)
end

local function redraw()
  vim.schedule(function ()
    vim.cmd.redraw()
  end)
end

api.nvim_create_autocmd('CmdlineEnter', {
  callback = function ()
    vim.ui_detach(ns)
    redraw()
  end
})

local qMsgs = {}

local function consumeQMsgs()
  for _, qmsg in ipairs(qMsgs) do
    log(vim.inspect(qmsg))
  end
  qMsgs = {}
end

local function attach()
  vim.ui_attach(ns, {ext_messages = true, ext_cmdline = false}, function (event, kind, content, replace_last)
    if event ~= 'msg_show' then
      return
    end
    local msg = {
      time = vim.uv.hrtime() / 1000000,
      fast = vim.in_fast_event(),
      e = event or 'none',
      k = kind or 'none',
      c = content or 'none',
      r = replace_last or 'none',
    }
    if not vim.in_fast_event() then
      consumeQMsgs()
      log(vim.inspect(msg))
      redraw()
      return
    end

    qMsgs[#qMsgs + 1] = msg
  end)
end

api.nvim_create_autocmd({'CmdlineLeave'}, {callback = attach})

api.nvim_create_autocmd({'UIEnter'}, {
  callback = function ()
    attach()
    vim.fn.inputlist({'Select: ', 1, 2})
  end
})
```

In short, I still don't understand when and how the queue should be consumed.

**Reply by luukvbaal** [reply]
> I suspect that consuming that queue along with the nearest message incoming from non-fast event is too late (at least in the same way as before).

Why do you think it is too late? Your `redraw()` is too late because it is scheduled. You should issue a non-deferred redraw whenever a prompt message happens (or in `cmdline_show` after #31525).

> Why not to display list_cmd above too? Not sure how it could affect external UIs.

What did you mean by this? What does above refer to?

**Reply by przepompownia** [reply]
> What did you mean by this? What does above refer to?

I meant the same effect like in built-in TUI (not sure if as the leading part of the prompt message)
![image](https://github.com/user-attachments/assets/3dd69aa1-37da-49e7-8a4d-13b3714b7bbf)

> Your redraw() is too late because it is scheduled. 

I overlooked that I did it in regardless of the case. Through trial and error I ended up with a version that (with #31525) is expected result (or close to it), because I see both messages at waiting for input:
![image](https://github.com/user-attachments/assets/bb1ca9d0-276b-4e76-a612-71fcadc42957)


```lua
local api = vim.api
local ns = api.nvim_create_namespace('messageRedirection')

local function log(msg)
  if 0 == #vim.trim(msg) then
    return
  end
  api.nvim_buf_set_lines(1, -1, -1, true, vim.split(msg, '\n'))
  vim.cmd.normal 'G'
end

api.nvim_create_autocmd('CmdlineEnter', {
  callback = function ()
    api.nvim__redraw({flush = true})
    vim.ui_detach(ns)
    api.nvim__redraw({flush = true})
  end
})

local qMsgs = {}

local function consumeQMsgs()
  for _, qmsg in ipairs(qMsgs) do
    log(vim.inspect(qmsg))
  end
  qMsgs = {}
end

local function attach()
  vim.ui_attach(ns, {ext_messages = true, ext_cmdline = false}, function (event, kind, content, replace_last)
    if event ~= 'msg_show' then
      return
    end
    if kind == 'return_prompt' then
      api.nvim_input('\r')
    end

    local msg = {
      time = vim.uv.hrtime() / 1000000,
      fast = vim.in_fast_event(),
      e = event or 'none',
      k = kind or 'none',
      c = content or 'none',
      r = replace_last or 'none',
    }
    if not vim.in_fast_event() then
      -- api.nvim__redraw({flush = true})
      consumeQMsgs()
      log(vim.inspect(msg))
      return
    end

    qMsgs[#qMsgs + 1] = msg
    vim.schedule(function ()
      consumeQMsgs()
      api.nvim__redraw({flush = true})
    end)
  end)
end

api.nvim_create_autocmd({'CmdlineLeave'}, {callback = attach})

api.nvim_create_autocmd({'UIEnter'}, {
  callback = function ()
    attach()
    -- vim.fn.inputlist({'Select: ', 1, 2})
  end
})
```

Notice `nvim__redraw` both before and after `ui_detach()`. I still haven't figured out what the minimum set of options I need to call `nvim__redraw()`. The commented call to `inputlist` on `UIEnter` still does not work as expected on my example.

**Reply by przepompownia** [reply]
@luukvbaal maybe still there are some details to fix (on the example or Nvim side) or discuss (like some mentioned above) here or in another place, but in general, thank you for the help and #31525. In summary, it's sufficient to display all expected contents before waiting for user input.

It looks ugly but works:
![image](https://github.com/user-attachments/assets/c636416e-6d4d-408c-a42f-9dc5092391aa)
